include/Contact.php

```diff
@@ -508,72 +508,81 @@ function contacts_not_grouped($uid,$start = 0,$count = 0) {
 /**
  * @brief Fetch the contact id for a given url and user
  *
+ * First lookup in the contact table to find a record matching either `url`, `nurl`,
+ * `addr` or `alias`.
+ *
+ * If there's no record and we aren't looking for a public contact, we quit.
+ * If there's one, we check that it isn't time to update the picture else we
+ * directly return the found contact id.
+ *
+ * Second, we probe the provided $url wether it's http://server.tld/profile or
+ * nick@server.tld. We quit if we can't get any info back.
+ *
+ * Third, we create the contact record if it doesn't exist
+ *
+ * Fourth, we update the existing record with the new data (avatar, alias, nick)
+ * if there's any updates
+ *
  * @param string $url Contact URL
- * @param integer $uid The user id for the contact
+ * @param integer $uid The user id for the contact (0 = public contact)
  * @param boolean $no_update Don't update the contact
  *
  * @return integer Contact ID
  */
 function get_contact($url, $uid = 0, $no_update = false) {
-	require_once("include/Scrape.php");
+	require_once "include/Scrape.php";
 
 	logger("Get contact data for url ".$url." and user ".$uid." - ".App::callstack(), LOGGER_DEBUG);;
 
 	$data = array();
-	$contactid = 0;
-
-	// is it an address in the format user@server.tld?
-	/// @todo use gcontact and/or the addr field for a lookup
-	if (!strstr($url, "http") OR strstr($url, "@")) {
-		$data = probe_url($url);
-		$url = $data["url"];
-		if ($url == "")
-			return 0;
-	}
-
-	$contact = q("SELECT `id`, `avatar-date` FROM `contact` WHERE `nurl` = '%s' AND `uid` = %d ORDER BY `id` LIMIT 2",
-			dbesc(normalise_link($url)),
-			intval($uid));
+	$contact_id = 0;
 
-	if (!$contact)
-		$contact = q("SELECT `id`, `avatar-date` FROM `contact` WHERE `alias` IN ('%s', '%s') AND `uid` = %d ORDER BY `id` LIMIT 1",
-				dbesc($url),
-				dbesc(normalise_link($url)),
-				intval($uid));
+	// Catch-all query, may return multiple rows
+	$contacts = q("SELECT `id`, `avatar-date` FROM `contact`
+				WHERE ('%s' IN (`url`, `addr`, `alias`) OR '%s' IN (`nurl`, `alias`))
```

Comment:
I ran a few tests, with a 5,000-rows `contact` table, an unsuccessful query is 0.05s, while a successful query is 0.01s. I expect this version to be definitely faster than the original when supplied a `nick@server.tld` URL as argument, and on par when supplied an existing `http://server.tld/nick` URL.